### ISSUE TITLE
[analyzer] False positive alpha.unix.cstring.UninitializedRead on pointer arithmetic across sub-array boundary

### ISSUE THREAD

**Issue by earnol** [llvm/llvm-project#199271]
After [#198346](https://github.com/llvm/llvm-project/pull/198346) `alpha.unix.cstring.UninitializedRead` reports a false positive when a pointer into a fully-initialized multidimensional array is advanced past the innermost dimension boundary and then used as a source argument to `memcpy`.

The array is fully zero-initialized with `= {}`, so all elements are defined. The pointer arithmetic crosses a sub-array boundary (which is technically UB per strict C24 standard reading of §6.5.7p8), but the memory is contiguous and initialized. The checker incorrectly reports the element as "undefined" rather than "out of bounds" — conflating sub-array boundary crossing with uninitialized memory.

## Reproducer
```c
typedef long a;
const a b[4][1200] = {};
void *memcpy(void *, const void *, unsigned long);
extern void *dst[];
void e() {
  const a *f = &b[0][0];
  int g = 2001;
  for (unsigned long i = 0; i < 2; i++) {
    memcpy(dst[i], f, sizeof(a));
    f = f + g;
  }
}
```

```
$ clang --analyze -Xclang -analyzer-checker=alpha.unix.cstring.UninitializedRead -Xclang -analyzer-checker=core -Xclang -analyzer-output=text test.c

test.c:9:5: warning: The first element of the 2nd argument is undefined [alpha.unix.cstring.UninitializedRead]
    memcpy(dst[i], f, sizeof(a));
    ^
```
Godbolt link available [here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:25,endLineNumber:6,positionColumn:25,positionLineNumber:6,selectionStartColumn:25,selectionStartLineNumber:6,startColumn:25,startLineNumber:6),source:'typedef+long+a%3B%0Aconst+a+b%5B4%5D%5B1200%5D+%3D+%7B%7D%3B%0Avoid+*memcpy(void+*,+const+void+*,+unsigned+long)%3B%0Aextern+void+*dst%5B%5D%3B%0Avoid+e()+%7B%0A++const+a+*f+%3D+%26b%5B0%5D%5B0%5D%3B%0A++int+g+%3D+2001%3B%0A++for+(unsigned+long+i+%3D+0%3B+i+%3C+2%3B+i%2B%2B)+%7B%0A++++memcpy(dst%5Bi%5D,+f,+sizeof(a))%3B%0A++++f+%3D+f+%2B+g%3B%0A++%7D%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:54.95194802148235,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'1',binary:'1',binaryObject:'0',commentOnly:'1',debugCalls:'1',demangle:'1',directives:'1',execute:'1',intel:'0',libraryCode:'1',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'--analyze+-Xclang+-analyzer-checker%3Dalpha.unix.cstring.UninitializedRead++-Xclang+-analyzer-checker%3Dcore+-Xclang+-analyzer-output%3Dtext',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'x86-64+clang+3.4+(assertions)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),k:45.04805197851764,l:'4',m:100,n:'0',o:'',s:1,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

The warning fires on the second loop iteration when `f` has been advanced by 2001 elements. The array `b` has innermost dimension 1200, so `f + 2001` crosses into `b[1][801]`. The total array size is `4 * 1200 = 4800` elements, so offset 2001 is well within the flat array bounds, and the entire array is zero-initialized.

## Expected behavior
No `UninitializedRead` warning. The memory at `f + 2001` is initialized (the entire array is `= {}`). 

If the analyzer wants to warn about pointer arithmetic crossing sub-array boundaries, it should use a different checker/message (e.g., `alpha.unix.cstring.OutOfBounds` or a new diagnostic), not `UninitializedRead`.

## Bisection
- **Fires**: after https://github.com/llvm/llvm-project/pull/198346 (commit `51279e7e0527`)
- **Silent**: before that commit

The PR fixed `CStringChecker::checkInit()` to correctly resolve buffer start offsets for offset buffer arguments. As a side effect, it now treats elements beyond the innermost sub-array dimension as "undefined" even when the containing array is fully initialized.

## Impact
This causes false positives in codebases that use pointer arithmetic on multidimensional arrays as flat buffers.

**Comment by NagyDonat** [comment]
@gamesh411 Could you check this issue?

**Comment by earnol** [comment]
I also noticed additional issue:
```c
typedef long a;
const a b[4][1200];
void *memcpy(void *, const void *, unsigned long);
extern void *dst[];
void e() {
  const a *f = &b[0][0];
  int g = 2001;
  for (unsigned long i = 0; i < 2; i++) {
    memcpy(dst[i], f, sizeof(a));
    f = f + g;
  }
}
```
Does not produce ANY diagnostics. But from language perspective ``b`` is global variable  so it is also zero initialized. Static analyzed does not account for it. But it's looks like an separate issue. Do you advice me to file the separate issue?

**Comment by gamesh411** [comment]
Hey @earnol!
Thanks for reporting this regression. I hope to handle it at the RegionStore level, but as a fallback, we can also fix the handling of multidimensional C arrays in the checker.
Still, I think the entire infrastructure would benefit from the fix in #200044.
